src/Microsoft.AspNet.OData.Shared/Builder/EdmModelHelperMethods.cs

```diff
@@ -638,13 +652,64 @@ private static void AddCoreVocabularyAnnotations(this EdmModel model, IEnumerabl
             if (edmProperties.Any())
             {
                 IEdmCollectionExpression collectionExpression = new EdmCollectionExpression(edmProperties.Select(p => new EdmPropertyPathExpression(p.Name)).ToArray());
-                IEdmTerm term = Microsoft.OData.Edm.Vocabularies.V1.CoreVocabularyModel.ConcurrencyTerm;
+                IEdmTerm term = CoreVocabularyModel.ConcurrencyTerm;
                 EdmVocabularyAnnotation annotation = new EdmVocabularyAnnotation(target, term, collectionExpression);
                 annotation.SetSerializationLocation(model, EdmVocabularyAnnotationSerializationLocation.Inline);
                 model.SetVocabularyAnnotation(annotation);
             }
         }
 
+        private static void AddDerivedTypeConstraintAnnotationForProperties(this EdmModel model, EdmTypeMap edmTypeMap)
+        {
+            Dictionary<IEdmProperty, PropertyConfiguration>
+                propertyConfiguration = edmTypeMap.EdmPropertyConfigurations;
+
+            foreach (var pair in propertyConfiguration)
+            {
+                if (pair.Value.DerivedTypeConstraints.Any())
+                {
+                    model.AddDerivedTypeConstraintAnnotation(pair.Key, edmTypeMap.EdmTypes, pair.Value.DerivedTypeConstraints.ToList());
+                }
+            }         
+        }
+
+        private static void AddDerivedTypeConstraintAnnotationForNavigationSource(this EdmModel model, IEdmVocabularyAnnotatable target,
+            NavigationSourceConfiguration navigationSourceConfiguration, EdmTypeMap edmTypeMap)
+        {
+            if (navigationSourceConfiguration != null)
+            {
+                model.AddDerivedTypeConstraintAnnotation(target,
+                    edmTypeMap.EdmTypes, navigationSourceConfiguration.DerivedTypeConstraints.ToList());
+            }
+        }
+
+        internal static void AddDerivedTypeConstraintAnnotation(this EdmModel model, IEdmVocabularyAnnotatable target, 
+            Dictionary<Type, IEdmType> edmTypeMap, List<Type> types)
+        {
+            IEdmTerm term = ValidationVocabularyModel.DerivedTypeConstraintTerm;
+            if (types.Count == 0)
+            {
+                return;
+            }
+
+            List<EdmStringConstant> collectionConstants = new List<EdmStringConstant>();
```

Comment:
>List<EdmStringConstant> collectionConstants = new List<EdmStringConstant>(); [](start = 12, length = 76)

shall we make sure no duplicated existed?

Reply:
We do that through the ISet. At this point in time, there should be no duplicates. 


---
In reply to: [358420621](https://github.com/OData/WebApi/pull/1977#discussion_r358420621) [](ancestors = 358420621)